src/main/java/dev/osm/mapsplit/ArrayMap.java

```diff
@@ -0,0 +1,145 @@
+package dev.osm.mapsplit;
+
+import static java.lang.Math.min;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * data structure that uses huge arrays with the OSM element's ID (the key) as the array index.
+ * 
+ * Because the number of IDs can (and in the case of nodes does) exceed Java's maximum array size,
+ * additional arrays are used for higher ids.
+ */
+public final class ArrayMap extends AbstractOsmMap {
+
+    /** maximum size for a single array, see e.g. https://stackoverflow.com/a/8381338/ */
+    private static final int MAX_ARRAY_SIZE = Integer.MAX_VALUE - 8;
+
+    private final long maxKey;
+
+    /** the basic array for IDs starting at 0. This is always used. */ 
+    private final @NotNull long[] array0;
+
+    /**
+     * additional arrays, only used if the maximum ID is large enough to require it.
+     * Not implemented as an array of arrays to avoid the additional indirection.
+     */
+    private final @Nullable long[] array1, array2, array3, array4, array5;
```

Comment:
I don't understand this. 

At best you are trading one shift (to determine the address of the array in the array holding the arrays) for a jump via a jump table (line 129) in which the jump target is determined by shifting too. IRL I doubt if there is a measurable performance impact one way or the other. But as a array of arrays implementation would avoid the repeated initialization code and hard coded size limit could be avoided, I would think this to much more preferable.

Reply:
You have a point, I've changed it to an array of arrays implementation in f82005d.